### PR TITLE
Enable Ollama integration test

### LINKED ISSUE #121: Improve Ollama IT by caching the image with model
Hi,

I have an improvement for [OllamaAutoConfigurationIT](https://github.com/spring-projects/spring-ai/blob/main/spring-ai-spring-boot-autoconfigure/src/test/java/org/springframework/ai/autoconfigure/ollama/OllamaAutoConfigurationIT.java) which can help when working locally. The current issue as described in the test is the model is around 3GB which can take quite a while depending on internet speed.

My approach using the Testcontainers lifecycle:
First test execution
1. Pull image from registry
2. Start container
3. Once the container has started, pull the model
4. Test execution
5. When test is finishing, commit the changes in the container and create a local image.
6. Stop container, stop test
Second test execution
1. Start container using the local image
2. Test execution
3. Stop container, stop test

If there is interest on this, I can provide the PR.

### PR DESCRIPTION
**Pull request by eddumelendez**
Currently, test is disable because in every run the image should
be downloaded and then pull the model. Now, taking advantage of
Testcontainers and a GHA, a new image is created on-the-fly with
the model in it and then cached, so, next executions will reuse the
cached image instead.

Fixes #121

### CODE REVIEW

File: spring-ai-spring-boot-autoconfigure/src/test/java/org/springframework/ai/autoconfigure/ollama/OllamaAutoConfigurationIT.java
Comment (markpollack) [inline spring-ai-spring-boot-autoconfigure/src/test/java/org/springframework/ai/autoconfigure/ollama/OllamaAutoConfigurationIT.java:1]: Thanks, I'd like to speed up the IT tests as well. I'm curious as to the rationale in going from a one liner to use test containers

```
	@Container
	static GenericContainer<?> ollamaContainer = new GenericContainer<>("ollama/ollama:0.1.23").withExposedPorts(11434);
```

To now using two static classes and a static method.  I'm not quite up on my testcontainer knowledge, but would have expected something more terse code-wise.

Comment (eddumelendez) [reply spring-ai-spring-boot-autoconfigure/src/test/java/org/springframework/ai/autoconfigure/ollama/OllamaAutoConfigurationIT.java:1]: The reason is to create the new Ollama image with the orca-mini model in it. We are doing it programmatically but another approach could be to reuse an existing image from Docker Hub that has the model and the one line could remain just pointing to the new image, removing the rest of the code. However, not sure if there is any preference to use any image out there or creating their own.

Comment (ThomasVitale) [reply spring-ai-spring-boot-autoconfigure/src/test/java/org/springframework/ai/autoconfigure/ollama/OllamaAutoConfigurationIT.java:1]: It's a tricky situation I encountered in other projects as well. On the one hand, it'd be better to use official images. On the other hand, downloading the model every time is not much feasible, requiring some mitigation strategies with additional configuration. I looked for existing images out there, but I ended up publishing my own Ollama-based images to make sure they are always up-to-date and for having multi-architecture support: https://github.com/ThomasVitale/llm-images.

**Comment by tzolov** [comment]
@eddumelendez , I'm trying to run this locally. The first attempt failed with 404 if not mistaken and a consecutive runs cause:
```
java.lang.ExceptionInInitializerError
 at java.base/java.util.ArrayList.forEach([ArrayList.java:1511](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
Caused by: org.testcontainers.containers.ContainerFetchException: Can't get Docker image: RemoteDockerImage(imageName=mistral-ollama/ollama:0.1.10, imagePullPolicy=org.springframework.ai.autoconfigure.ollama.OllamaChatAutoConfigurationIT$OllamaContainer$$Lambda$497/0x000000013a226ac8@1e7f2e0f, imageNameSubstitutor=org.testcontainers.utility.ImageNameSubstitutor$LogWrappedImageNameSubstitutor@1da6ee17)
 at org.testcontainers.containers.GenericContainer.getDockerImageName([GenericContainer.java:1364](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 at org.testcontainers.containers.GenericContainer.doStart([GenericContainer.java:359](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 at org.testcontainers.containers.GenericContainer.start([GenericContainer.java:330](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 at org.springframework.ai.autoconfigure.ollama.OllamaChatAutoConfigurationIT.([OllamaChatAutoConfigurationIT.java:69](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html))
 ... 2 more
Caused by: com.github.dockerjava.api.exception.NotFoundException: Status 404: {"message":"pull access denied for mistral-ollama/ollama, repository does not exist or may require 'docker login': denied: requested access to the resource is denied"}
```

May you know what is going on?

**Comment by tzolov** [comment]
Did a second attempt (after removing all docker images) and the error it fails with is :
```
8:37:40.513 [main] INFO tc.ollama/ollama:0.1.10 -- Image ollama/ollama:0.1.10 pull took PT29.724702S
18:37:40.513 [docker-java-stream-1404029602] INFO tc.ollama/ollama:0.1.10 -- Pull complete. 3 layers, pulled in 28s (downloaded 471 MB at 16 MB/s)
18:37:40.518 [main] INFO tc.ollama/ollama:0.1.10 -- Creating container for image: ollama/ollama:0.1.10
18:37:44.108 [main] INFO tc.ollama/ollama:0.1.10 -- Container ollama/ollama:0.1.10 is starting: 8aff3a52c42c029a229503dc4e50f59e80ee5b1074734df00852b9bd01de8af5
18:37:44.452 [main] INFO tc.ollama/ollama:0.1.10 -- Container ollama/ollama:0.1.10 started in PT3.934304S
18:39:38.730 [main] INFO org.springframework.ai.autoconfigure.ollama.OllamaChatAutoConfigurationIT -- Start pulling the 'mistral ' generative ... would take several minutes ...
18:43:47.751 [main] INFO org.springframework.ai.autoconfigure.ollama.OllamaChatAutoConfigurationIT -- mistral pulling competed!
18:43:49.397 [main] WARN org.springframework.ai.ollama.api.OllamaApi -- [404] Not Found - 404 page not found
```

**Comment by eddumelendez** [comment]
Hi @tzolov, I've reviewed and fixed the issue. PR is updated and everything should work as expected now.

**Comment by tzolov** [comment]
Thanks @eddumelendez , just merged it but it seems we don't have the right account for using docker caching ? https://github.com/spring-projects/spring-ai/actions/runs/8070608896
Or is there a different way to resolve this?

**Comment by eddumelendez** [comment]
Looks like need to check `Actions Permissions` under `Settings > Actions > General` to allow external actions.

**Comment by tzolov** [comment]
Yeh,  got this as well. But i'm not sure how safe is to white list an unverified action?

**Comment by tzolov** [comment]
@eddumelendez i've whitelisted it but now it fails on this
Screenshot 2024-02-27 at 22 55 22

**Comment by eddumelendez** [comment]
Let me try on my fork.

> But i'm not sure how safe is to white list an unverified action?

Totally understand. Another alternative could be have persistent runners

**Comment by eddumelendez** [comment]
Couldn't reproduce it https://github.com/eddumelendez/spring-ai/actions/runs/8072468166/job/22054281356#step:8:22

Can you enable debug logs, please?

**Comment by izeye** [comment]
This seems to have been merged in 246ba173fec3ee135bed1d30f2c64117e69e3f28.

**Comment by tzolov** [comment]
@izeye it is merged but disabled because of the https://github.com/spring-projects/spring-ai/pull/322#issuecomment-1967677261 we still have to resolve.

File: spring-ai-spring-boot-autoconfigure/src/test/java/org/springframework/ai/autoconfigure/ollama/OllamaAutoConfigurationIT.java
Comment (markpollack) [reply spring-ai-spring-boot-autoconfigure/src/test/java/org/springframework/ai/autoconfigure/ollama/OllamaAutoConfigurationIT.java:1]: That image has been closed with 'won't do'  - so closing this issue.  We will have to setup a separate integration test profile for ollama along with other model providers that aren't currently running in GI

**Comment by markpollack** [comment]
Closing as model specific images are not going to be provided by ollama.

We will pick this up as part of a larger effort to have more CI coverage outside the current github action that runs on each commit.

**Comment by dsyer** [comment]
FWIW I just tried my demo in a github action and the ollama models (x2) pulled in 30sec  It's almost not worth trying to cache. If only my network was that fast.

I also tried using https://docs.docker.com/build/ci/github-actions/cache/ to see if it would help, but it fails to create the cache because (I think) ollama is running as root so the file permissions are fubar.

**Comment by dsyer** [comment]
Update. I got it working with this

```java
		@Bean
		@ServiceConnection
		public ChromaDBContainer chroma() {
			return new ChromaDBContainer(DockerImageName.parse("ghcr.io/chroma-core/chroma:0.5.5"));
		}

@Bean
@ServiceConnection
public OllamaContainer ollama() throws Exception {
	@SuppressWarnings("resource")
	OllamaContainer ollama = new OllamaContainer(DockerImageName.parse("ollama/ollama:0.3.2"))
			// Not recommended strategy from testcontainers, but the only practical way to
			// make it work locally
			.withFileSystemBind("ollama", "/root/.ollama", BindMode.READ_WRITE);
	return ollama;
}
@Bean
ApplicationRunner runner(OllamaContainer ollama) {
	return args -> {
		logger.info("Pulling models...");
		ollama.execInContainer("ollama", "pull", "albertogg/multi-qa-minilm-l6-cos-v1");
		ollama.execInContainer("ollama", "pull", "mistral");
		ollama.execInContainer("chmod", "go+r", "-R", "/root/.ollama");
		logger.info("...done");
	};
}
```

but creating the cache (on the first run) and unpacking it (subsequently) takes about 1 minute. So it's not efficient for this case. Maybe it would work better with different models or more models.

**Comment by ThomasVitale** [comment]
We might be able to close this issue now. All integration tests for Spring AI Ollama are now enabled and running on Testcontainers, also on the GitHub Actions workflow. They are based on the newly introduced "model auto-pull" feature https://github.com/spring-projects/spring-ai/pull/1554 It's used for the integration test setup but it's also a feature to improve the developer experience, making it possible to pull models automatically at startup time if not available yet.

**Comment by markpollack** [comment]
I don't see the ollama tests running in the CI log, still something to sort out.  

@dsyer thanks for taking the time to dig into it.  We do pull several models in the range of IT tests ,but that info you provided is very useful.  Let's see how it goes.

I'd like to move this issue to https://github.com/spring-projects/spring-ai-integration-tests/issues/5 as we now have a more dedicated environment to run all the tests, thus decreasing the already considerable time spent for our mainline build.

**Pull request closed without merging**